Added custom exceptions.

Srujan-D opened the pull request:
1) Codes for the exceptions are in (gennav.custom_exceptions.utils)
2) Exceptions added for:
    1. Key_Error- custom messege will be generated when key is not found in the dictionary containing all the parameters.
    2. Start state is in obstacle.
    3. End/Goal state is in obstacle.
    4. Path is not found by the planner.
    5. Sampler is unable to sample a state.
3) These exceptions have been integrated with all the path planners.
4) Generated docs have been added.

Srujan-D committed:
Added custom exceptions.
1) Codes for the exceptions are in (gennav.custom_exceptions.utils)
2) Exceptions added for:
    1. Key_Error- custom messege will be generated when key is not found in the dictionary containing all the parameters.
    2. Start state is in obstacle.
    3. End/Goal state is in obstacle.
    4. Path is not found by the planner.
    5. Sampler is unable to sample a state.
3) These exceptions have been integrated with all the path planners.
4) Generated docs have been added.

Srujan-D commented:
1) In custom exceptions file, try/except should work for py2 and py3 right? Its still throwing a syntax error (see travis ci logs for py2). 
2) I don't know why i'm getting the lint error for incorrect sorting in polygon_env. It isn't showing any message when I do it locally.
4) I have added Key_Error custom exception right now. But need to discuss before using it (I'll need to make appropriate changes in all the plan methods to integrate it).
3) @sushant1212  I think there's some bug in informed_rrtstar.py. It's returning a single state path sometimes.

sushant1212 commented:
>     1. In custom exceptions file, try/except should work for py2 and py3 right? Its still throwing a syntax error (see travis ci logs for py2).
> 
>     2. I don't know why i'm getting the lint error for incorrect sorting in polygon_env. It isn't showing any message when I do it locally.
> 
>     3. I have added Key_Error custom exception right now. But need to discuss before using it (I'll need to make appropriate changes in all the plan methods to integrate it).
> 
>     4. @sushant1212  I think there's some bug in informed_rrtstar.py. It's returning a single state path sometimes.

Whenever solution is not found, then it returns only the start point. That's why you're getting single state paths. Atharv told that we would replace it later with PathNotFound error

Srujan-D commented:
> Whenever solution is not found, then it returns only the start point. That's why you're getting single state paths. Atharv told that we would replace it later with PathNotFound error

But even when this error is not raised, the path visualized is similar to this:
![Screenshot from 2020-08-02 22-39-49](https://user-images.githubusercontent.com/53463585/89155390-59f84c00-d586-11ea-8fb1-0459d76d4bf8.png)

sushant1212 commented:
> > Whenever solution is not found, then it returns only the start point. That's why you're getting single state paths. Atharv told that we would replace it later with PathNotFound error
> 
> But even when this error is not raised, the path visualized is similar to this:
> ![Screenshot from 2020-08-02 22-39-49](https://user-images.githubusercontent.com/53463585/89155390-59f84c00-d586-11ea-8fb1-0459d76d4bf8.png)

Ok I'll look into it

threewisemonkeys-as commented:
1. You are using f-strings which arent supported by y2
2. Is tox running ok locally for lint?
3. No need for the key error as we discussed, it can be removed

Srujan-D commented:
>     1. You are using f-strings which arent supported by y2

>     2. Is tox running ok locally for lint?
              
>     3. No need for the key error as we discussed, it can be removed
              
1. Yeah, but that's why I'm doing `try:` / `except SyntaxError:` . I thought that will work, but it isn't. I can ofc just have `return "%s"%(var)`, but idk if we should have that for py3 (doesn't give an error, but its old).
2. Nvm, lint errors are sorted.
3. Yeah, I'll remove that.

threewisemonkeys-as commented:
> Yeah, but that's why I'm doing try: / except SyntaxError: . I thought that will work, but it isn't. I can ofc just have return "%s"%(var), but idk if we should have that for py3 (doesn't give an error, but its old). 

Stick to the old format. We should avoid using try accept for python versions wherever possible

threewisemonkeys-as commented on gennav/utils/custom_exceptions.py:
```diff
@@ -0,0 +1,99 @@
+class Error(Exception):
```
Is this neccessary, cant they all inherit `Exception`  directly

Srujan-D replied:
Not necessary. But in gen, almost in all custom exceptions examples (including the python docs), there's a base Error class similar to this. Should I remove it?

threewisemonkeys-as replied:
Yeah, it might be useful if we abstract things into the base class. For now you can remove it

Srujan-D replied:
Okay.

threewisemonkeys-as commented on gennav/utils/custom_exceptions.py:
```diff
@@ -0,0 +1,99 @@
+            return f"{self.key} key not found in the dictionary containing all parameters ({self.dict.keys()})."
+
+
+class StartStateinObs(Error):
```
could you name this `InvalidStateState`

Srujan-D replied:
Yeah.

threewisemonkeys-as commented on gennav/utils/custom_exceptions.py:
```diff
@@ -0,0 +1,99 @@
+            return f"{self.start} -> {self.message}"
+
+
+class GoalStateinObs(Error):
```
Similar name change here to `InvalidGoalState`

Srujan-D replied:
Yeah.

Srujan-D committed:
Made changes to custom_exceptions.
1) Moved custom exceptions from gennav.utils to gennav.utils.exceptions .
2) Changed the return statement of all the custom_exceptions to suit py2 as well as py3.
3) Renamed GoalStateinObs and StartStateinObs.
4) Made necessary changes to all file importing custom_exceptions.
5) Removed Key_Error exception from custom_exceptions.
6) Fixed lint errors.
7) (Minor): Changed buffer_dist in rrt_test file.

Srujan-D commented:
Also, getting an `ImportError` in py2 after moving custom exceptions from `.utils` to `.utils.exceptions`.

threewisemonkeys-as commented:
Try to use absolute imports instead of relative ones

Srujan-D commented:
I have used absolute imports only, forgot to add gennav.utils.... here. Also, the ImportError is for py2 only. For py3, its working fine.

threewisemonkeys-as commented:
Actuallly, you can keept the excpetions in `utils/exception.py` and add `import gennav.utils.exceptions` in `utils/__init__.py`

Srujan-D commented:
Cool, I'll do that.

Srujan-D commented:
> Actuallly, you can keept the excpetions in `utils/exception.py` and add `import gennav.utils.exceptions` in `utils/__init__.py`

It still gives this import error: 
`gennav/utils/__init__.py:2: in <module>`
`import gennav.utils.exceptions.custom_exceptions  # noqa: F401`
`E   ImportError: No module named exceptions.custom_exceptions`

Srujan-D committed:
Removed base Error class from custom_exceptions
1) Also:
    1. Removed custom_exceptions from utils.exceptions to utils.
    2. Made necessary changes everywhere.
    3. Added import lione for custom_exceptions in gennav.utils.__init__
2) Added PathNotFound Error in A* (gennav.utils.graph_search)

Srujan-D commented:
> Actuallly, you can keept the excpetions in `utils/exception.py` and add `import gennav.utils.exceptions` in `utils/__init__.py`

Sort, the only fail in the build is for the I-RRT* which sushant resolved in his PR i think.

Srujan-D committed:
Fixed lgtm alert
